### PR TITLE
Add a way to inject theme type, add sc extension test, allow extending with classes and SFCs

### PR DESCRIPTION
**Pull request by eloiqs**
Those are improvment suggestions for your PR (https://github.com/DefinitelyTyped/DefinitelyTyped/pull/29549)

### CODE REVIEW

File: types/rebass/index.d.ts [hunk 1]
```diff
@@ -66,7 +66,7 @@ export interface BoxProps
         WidthProps,
         FontSizeProps,
         RebassHTMLAttributes<HTMLDivElement> {}
-type BoxClass = StyledComponentClass<BoxProps, {}>
+type BoxClass = React.StatelessComponent<ThemedOuterStyledProps<BoxProps, {}>>
```
Comment (alexdriaguine) [inline types/rebass/index.d.ts:69]: Why the change from `StyledComponentClass`? 🤔

File: types/rebass/rebass-tests.tsx [hunk 2]
```diff
@@ -267,3 +271,81 @@ const AsProp = () => (
...
+interface Theme {
+    dark: {
+        backgroundColor: string
+    }
+}
+
+const WithTypedTheme = () => {
+    const {Box} = rebassComponents as RebassComponentsModule<Theme>
+    return (
+        <Box theme={{dark: {backgroundColor: '#666666'}}} />
```
Comment (alexdriaguine) [inline types/rebass/rebass-tests.tsx]: Hmm.. Not sure i follow the use case here. According to styled components documentation, theming is done by using a `<ThemeProvider theme={theme}>...</ThemeProvider>`, not sure why one would want to one-off theme components like this. You got some example use case 😁?. Check it out https://rebassjs.org/theming

**Review by alexdriaguine — CHANGES_REQUESTED** [review]
I think this looks good, I'm just a bit confused about the `RebassComponentsModule` export and passing a single `theme` prop to components usage 😕 

Here's my reasoning about theming rebass components:

Theming is done via a top level `ThemeProvider`. If we would like to have dark/light themes, one would do something like this:
```typescript
class CustomThemeProvider extends Component {
  state = {theme: "dark"}
  changeTheme = (theme: string) => this.setState({theme})
  render() { 
    return <ThemeProvider theme={this.state.theme}>
	  <ThemeToggler changeTheme={this.changeTheme}
	  {children}
	</ThemeProvider>
  }
}
```

Theming one off components such as button for different usages is done via the `variant` prop and suppliying that theme to the `ThemeProvider`.

Extending the components looks good :D

**Comment by alexdriaguine** [comment]
Nevermind the `passing a single theme prop to components usage 😕` comment, I just read the SC docs and saw that this is a legit use case (now I learned something new today! yay) 😁

**Comment by alexdriaguine** [comment]
So i think I get `RebassComponentsModule`, its so you get a typed `theme` prop then doing `<Box theme=.../>`, right? Wouldn't a better case be to export these components directly with a `Theme` argument?

```typescript
interface RebassTheme {
 //.. all rebass theme keys
}

export Box: StyledComponentClass<BoxProps, RebassTheme>
```

and when you want to extend the theme to include stuf like `dark`,

```typescript
interface MyTheme extends RebassTheme {
  dark: any
}


const DarkBox = () => {
	const CustomBox = Box as StyledComponentsClass<BoxProps, MyTheme>
    return <CustomBox theme={{dark: ... }} />
}
```

(Not tested code 😅 )

File: types/rebass/index.d.ts (hunk 1 above)
Comment (eloiqs) [reply types/rebass/index.d.ts:69]: Because a `StyledComponentClass` has a `withComponent` method and rebass components have not. 

If you take a look at `StyledComponentClass` you'll notice it extends `React.ComponentClass<ThemedOuterStyledProps<O, T>>` which, when instantiated, returns a `React.Component<ThemedOuterStyledProps<O, T>>`. Moreover, I don't expect people to use any of the `React.ComponentClass` attributes, if I'm not mistaken, those would be for setting propTypes, defaultProps, etc. on rebass components which would affect all instances. I am not seeing the use case for doing that.

I also favored using `React.StatelessComponent` instead of `React.Component` because I don't think we need the `setState`, `forceUpdate`, etc. methods on rebass components.

**Comment by eloiqs** [comment]
If i'm not mistaken, rebass doesn't ship with a default theme, so `RebassTheme` would just be `{}` right?

Your way to inject the theme type works, you can still do it like that if you want.

My rationale for creating the `RebassComponentsModule<T>` was for allowing people to only have to do it once. I was inspired by the style-components typings (https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/styled-components/index.d.ts#L149) .

With this interface you could create a file like `theme.ts`, define your theme, import all the rebass components and re-export like so: 
```ts
import * as rebass from 'rebass'
import * as sc from 'styled-components'

export const theme = {
	backgroundColor: 'black'
}

const { Box, Heading, Card } = rebass as rebass.RebassComponentsModule<typeof Theme>
const { default: styled } = sc as sc.ThemedStyledComponentsModule<typeof Theme>

export {Box, Heading, Card, styled}
```

Then in you component files you would only have to 
```ts
import { ThemeProvider } from 'styled-components'
import { theme, Box, styled } from './theme'

const Container = styled(Box)`
    background-color: ${props => props.theme.backgroundColor};
    width: 75%;
    margin: 'auto';
`

export const App = (props) => (
    <ThemeProvider theme={theme}>
        <Box>
            {props.children}
        </Box>
    </ThemeProvider>
)
```

**Comment by alexdriaguine** [comment]
Ah, thanks for the clarification, now it makes sense! 

As for the type of components, it seems like the withComponent might be deprecated in the future since the `as` prop. And since rebass is also supposed to work with emotion I think it makes sense to have components be typed as `React.StatelessComponent` 😁
